cpp/src/cylon/util/arrow_utils.cpp

```diff
@@ -275,32 +279,79 @@ arrow::Status SampleArray(const std::shared_ptr<arrow::ChunkedArray> &arr,
-    case arrow::Type::INT32:return sample_fixed_size_array<arrow::Int32Type>(arr, num_samples, out, pool);
-    case arrow::Type::UINT64:return sample_fixed_size_array<arrow::UInt32Type>(arr, num_samples, out, pool);
-    case arrow::Type::INT64:return sample_fixed_size_array<arrow::Int64Type>(arr, num_samples, out, pool);
-    case arrow::Type::FLOAT:return sample_fixed_size_array<arrow::FloatType>(arr, num_samples, out, pool);
-    case arrow::Type::DOUBLE:return sample_fixed_size_array<arrow::DoubleType>(arr, num_samples, out, pool);
-    case arrow::Type::DATE32:return sample_fixed_size_array<arrow::Date32Type>(arr, num_samples, out, pool);
-    case arrow::Type::DATE64:return sample_fixed_size_array<arrow::Date64Type>(arr, num_samples, out, pool);
-    case arrow::Type::TIMESTAMP:return sample_fixed_size_array<arrow::TimestampType>(arr, num_samples, out, pool);
-    case arrow::Type::TIME32:return sample_fixed_size_array<arrow::Time32Type>(arr, num_samples, out, pool);
-    case arrow::Type::TIME64:return sample_fixed_size_array<arrow::Time64Type>(arr, num_samples, out, pool);
-    case arrow::Type::STRING: return sample_binary_array<arrow::StringType>(arr, num_samples, out, pool);
-    case arrow::Type::BINARY:return sample_binary_array<arrow::BinaryType>(arr, num_samples, out, pool);
-    case arrow::Type::FIXED_SIZE_BINARY:
-      return sample_fixed_size_array<arrow::FixedSizeBinaryType>(arr,
-                                                                 num_samples,
-                                                                 out,
-                                                                 pool);
-    default:return arrow::Status(arrow::StatusCode::Invalid, "unsupported type");
-  }
+
+    // if num_samples == 0, just finish the array and return OK
+    if (num_samples == 0) {
+        out = GetChunkOrEmptyArray(arr, 0, pool);
+        return arrow::Status::OK();
+    }
+
+    // if the num_samples is greater than the array length, don't sample, just return the array
+    if ((int64_t) num_samples >= arr->length()) {
+        if (arr->num_chunks() > 1) {
+            const auto &res = arrow::Concatenate(arr->chunks(), pool);
+            RETURN_ARROW_STATUS_IF_FAILED(res.status());
+            out = res.ValueOrDie();
+        } else {
+            out = arr->chunk(0);
+        }
+        return arrow::Status::OK();
+    }
+
+    // general case
+    static std::random_device rd;
+    static std::mt19937_64 gen(rd());
+    std::uniform_int_distribution<int64_t> distrib(0, arr->length() - 1);
+
+    arrow::Int64Builder builder(pool);
+    RETURN_ARROW_STATUS_IF_FAILED(builder.Reserve((int64_t) num_samples));
+    for (uint64_t i = 0; i < num_samples; i++) {
+        builder.UnsafeAppend(distrib(gen));
+    }
+
+    auto arr_result = builder.Finish();
+
+    RETURN_ARROW_STATUS_IF_FAILED(arr_result.status());
+
+    auto indices = arr_result.ValueOrDie();
```

Comment:
```suggestion
ARROW_ASSIGN_OR_RAISE(auto indices, builder.Finish());
```
This is a convenient macro where `arrow::Result` is returned.

Reply:
this should be done